pkg/registry/apis/query/README.md

```diff
@@ -0,0 +1,77 @@
+# Query service
+
+This query service aims to replace the existing /api/ds/query.  
+
+The key differences are:
+1. This service has a stronger type system (not simplejson)
+2. Same workflow regardless if expressions exist
+3. Datasource settings+access is managed in each datasource, not at the beginning
+
+
+
+### Current /api/ds/query workflow
+
+```mermaid
+sequenceDiagram
+    autonumber
+    actor User as User or Process
+    participant api as /api/ds/query
+    participant db as Storage<br/>(SQL)
+    participant ds as Datasource<br/>Plugin
+    participant expr as Expression<br/>Engine
+
+    User->>api: POST Query
+    loop Each query
+        api->>api: Parse query
+        api->>db: Get ds config<br>and secrets
+        db->>api: 
+    end
+    alt No expressions
+      alt Single datasource
+          api->>ds: QueryData
+      else Multiple datasources
+        loop Each datasource
+          api->>ds: QueryData
+        end
+        api->>api: Wait for results
+      end
+    else Expressions exist
+        api->>expr: Calculate expressions graph
+        loop Each node (eg, refID)
+          alt Is query
+              expr->>ds: QueryData
+          else Is expression
+            expr->>expr: Process
+          end
+        end
+    end
+    api->>User: return results
+```
+
+
+
+### /apis/query.grafana.app (in single tenant grafana)
+
+```mermaid
+sequenceDiagram
+    autonumber
+    actor User as User or Process
+    participant api as /apis/query.grafana.app
+    participant ds as Datasource<br/>Handler/Plugin
+    participant db as Storage<br/>(SQL)
+    participant expr as Expression<br/>Engine
+
+    User->>api: POST Query
```

Comment:
This might be controversial and I've raised it before. But wouldn't having one query endpoint and one expr endpoint make sense? I think so. It would make the intent clearer from the consumer, allow request metrics to capture this intent and to compare the endpoints and allow having potential different SLO's of the endpoints. You could also streamline the request model to better/easier provide a model that's easier to construct and digest when expressions is in the mix. 

Part of this discussion is whether query and SSE should be the same service or different ones. Right now they're kind of both which doesn't feel optimal. Thinking of ownership and such as well moving forward.

Reply:
@kylebrandt do you have any insights/thoughts about this?

Reply:
That all makes sense -- and for sure something we should explore when designing /v2 api for this.  Right now this /query endpoint aims to be a drop in replacement for the /api/ds/query path.

With apisevers for each datasource, we do have clear place for SLOs and ownership of the sub-requests.  The generic /query endpoint could be limited to queries that include expressions or mixed datasources